{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Purpose: pipeline so that given a list of CCSs, for each label where the 5p barcode, 3p barcode, and adapter lie "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [],
   "source": [
    "import CCS_Filter as cf\n",
    "import numpy as np\n",
    "import random as rd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adaptation of CCS_Filter's score alignment, except this time \n",
    "#RETURNS list of score, traceback alignment (just reference portion)\n",
    "#FOR NOW: linear gap penalty d\n",
    "#parameters: gap penalty d, barcode/reference(vertical), seq y(horizontal),\n",
    "    #score matrix, reference character\n",
    "#reference character: when returning alignment, replace all characters in seq y\n",
    "    # with the reference character. Then only return the line of the alignment for \n",
    "    # the reference\n",
    "    #example of return: 55555--5----------------\n",
    "    #if the reference is the 5p barcode\n",
    "#initialization: 0s in first row; (i,0) = -i*d\n",
    "#Traceback favors iy>m>ix becuase more likely to have gaps in ref than \n",
    "    #y since reference shorter than y\n",
    "    \n",
    "\n",
    "def glocal_alignment(gap_penalty, sequence, reference, score_dict, reference_char):\n",
    "    gap_penalty = np.abs(gap_penalty)\n",
    "    #matrix\n",
    "    num_rows = len(reference) +1\n",
    "    num_cols = len(sequence) +1\n",
    "    F = np.zeros((num_rows, num_cols))\n",
    "    T = np.zeros((num_rows,num_cols)) #traceback\n",
    "    #labels for tracing, prioritizing iy>m>ix\n",
    "    IY, M, IX = 3, 2, 1\n",
    "    T_list = [IY, M, IX]\n",
    "    #initialization\",\n",
    "    #row 0 already initialized to zeros\n",
    "    #keep T[0][j] as zeros because can stop anywhere there\n",
    "    for i in np.arange(1,num_rows):\n",
    "        F[i][0] = -1*i*gap_penalty\n",
    "        T[i][0] = IX\n",
    "        \n",
    "\n",
    "    for i in np.arange(1,num_rows): # i represents the row\n",
    "        for j in np.arange(1,num_cols): #j represents the col\n",
    "#            print('i is', i)\n",
    "#            print('j is', j)\n",
    "#            print(\"looking at ref(x) char\", reference[i-1])\n",
    "#            print(\"looking at seq(y) char\", sequence[j-1])\n",
    "            m = F[i-1][j-1] + score_dict[(reference[i-1],sequence[j-1])]\n",
    "            ix = F[i-1][j] - gap_penalty\n",
    "            iy = F[i][j-1] - gap_penalty\n",
    "#            print('m is', m)\n",
    "#            print('ix is',ix)\n",
    "#            print('iy is', iy)\n",
    "            F[i][j], T[i][j] = max(list(zip([iy,m,ix],T_list)))\n",
    "#            print(\"T entry is: \",T[i][j])\n",
    "    #now matrix is filled\"\n",
    "    last_row_array = np.array(F[num_rows-1])\n",
    "    max_indices = np.where(last_row_array == max(F[num_rows-1]))\n",
    "    max_indices_list = []\n",
    "    for elem in max_indices:\n",
    "        for entry in elem:\n",
    "            max_indices_list.append(entry)\n",
    "#    print('max_indices is:',max_indices_list)\n",
    "\n",
    "\n",
    "    score_list = [F[num_rows-1][i] for i in max_indices_list]\n",
    "#    print(\"T is: \")\n",
    "#    print(T)\n",
    "    x_alignments = []\n",
    "    y_alignments = []\n",
    "    for trace_start_col in max_indices_list:\n",
    "        #start traceback\n",
    "        x_align = [] #reference\n",
    "        y_align = [] #seq\n",
    "        seq_length = len(sequence)\n",
    "        y = sequence\n",
    "        x = [reference_char]*len(reference)\n",
    "        i_fin, j_fin = num_rows-1, trace_start_col\n",
    "        i = i_fin\n",
    "        j = j_fin\n",
    "        traceback = T[i][j]\n",
    "        current = ''\n",
    "        if j_fin < seq_length:\n",
    "            for n in np.arange(seq_length - j_fin):\n",
    "                x_align.append('-')\n",
    "                y_align.append(y[seq_length-n-1])\n",
    "        if traceback == M:\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "            y_align.append(y[j])\n",
    "            x_align.append(x[i])\n",
    "        elif current == IY:\n",
    "            j -= 1\n",
    "            x_align.append('-')\n",
    "            y_align.append(y[j])\n",
    "        else: #current == IX\n",
    "            i -= 1\n",
    "            x_align.append(x[i])\n",
    "            y_align.append('-')\n",
    "\n",
    "\n",
    "        current = T[i][j]\n",
    "\n",
    "        while i>0 and j>0:\n",
    "            if current == M:\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "                y_align.append(y[j])\n",
    "                x_align.append(x[i])\n",
    "            elif current == IY:\n",
    "                j -= 1\n",
    "                x_align.append('-')\n",
    "                y_align.append(y[j])\n",
    "            else: #current == IX\n",
    "                i -= 1\n",
    "                x_align.append(x[i])\n",
    "                y_align.append('-')\n",
    "            current = T[i][j]\n",
    "\n",
    "        if j >0:\n",
    "            while j >0:\n",
    "                j -= 1\n",
    "                x_align.append('-')\n",
    "                y_align.append(y[j])\n",
    "\n",
    "\n",
    "        x_align.reverse()\n",
    "        y_align.reverse()\n",
    "        x_alignments.append(x_align)\n",
    "        y_alignments.append(y_align)\n",
    "        \n",
    "    return_list = list(zip(score_list,x_alignments))\n",
    "    \n",
    "#    x_align_str = \"\".join(x_align)\n",
    " #   y_align_str = \"\".join(y_align)\n",
    "    \n",
    "    \n",
    "#    print('y(seq): ', y_align)\n",
    " #   print('x(ref): ', x_align)\n",
    " #   print('Returning: ',return_list)\n",
    "    return return_list\n",
    "reference = \"ACGT\"\n",
    "ref_rc = cf.gen_rev_complement(reference)\n",
    "seq_0 = \"ACGATAGA\"\n",
    "seq0_rc = cf.gen_rev_complement(seq_0)\n",
    "seq_1 = \"AAGACAGT\"\n",
    "seq1_rc = cf.gen_rev_complement(seq_1)\n",
    "seq_2 = \"AGGACGCTAA\"\n",
    "seq2_rc = cf.gen_rev_complement(seq_2)\n",
    "glocal_alignment(gap_penalty=1, sequence=test_3C01_strings[0], reference=three_adapter, score_dict=cf.score_dict, reference_char = 'R')\n",
    "a=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len two_adapter: 45\n",
      "len three_adapter: 45\n",
      "len four_adapter: 45\n",
      "len 5p barcode: 25\n",
      "len barcode2: 71\n",
      "len barcode3: 71\n",
      "len barcode4: 71\n",
      "len two_adapter + 3p barcode: 116\n",
      "len three_adapter + 5p barcode: 70\n",
      "len four_adapter + 5p barcode: 70\n",
      "total:  141\n"
     ]
    }
   ],
   "source": [
    "#test strings\n",
    "nucleotides = cf.nucleotides\n",
    "\n",
    "fivePBarcode = cf.barcodes_list[0]\n",
    "fivePBarcodeRC = cf.gen_rev_complement(fivePBarcode)\n",
    "barcode2 = cf.barcode_2\n",
    "barcode3 = cf.barcode_3\n",
    "barcode4 = cf.barcode_4\n",
    "barcode2RC = cf.barcode_2_rc\n",
    "barcode3RC = cf.gen_rev_complement(barcode3)\n",
    "barcode4RC = cf.gen_rev_complement(barcode4)\n",
    "#adapter\n",
    "two_adapter = ''\n",
    "three_adapter = ''\n",
    "four_adapter = ''\n",
    "#/tier2/deweylab/data/thomson_lab/Pacbio/2_B01/m54178_180915_120213.adapters.fasta\n",
    "with open('two_B01.adapters.fasta') as f: \n",
    "    for line in f: \n",
    "        line = line.strip()\n",
    "        if line[0]!='>':\n",
    "            two_adapter = line\n",
    "\n",
    "#/tier2/deweylab/data/thomson_lab/Pacbio/3_C01/m54178_180916_082219.adapters.fasta\n",
    "with open('three_C01.adapters.fasta') as f: \n",
    "    for line in f: \n",
    "        line = line.strip()\n",
    "        if line[0]!='>':\n",
    "            three_adapter = line\n",
    "\n",
    "#/tier2/deweylab/data/thomson_lab/Pacbio/4_D01/m54178_180917_044252.adapters.fasta            \n",
    "with open('four_D01.adapters.fasta') as f: \n",
    "    for line in f: \n",
    "        line = line.strip()\n",
    "        if line[0]!='>':\n",
    "            four_adapter = line\n",
    "two_adapter_RC = cf.gen_rev_complement(two_adapter)\n",
    "three_adapter_RC = cf.gen_rev_complement(three_adapter)\n",
    "four_adapter_RC = cf.gen_rev_complement(four_adapter)\n",
    "print(\"len two_adapter:\", len(two_adapter))\n",
    "print(\"len three_adapter:\", len(three_adapter))\n",
    "print(\"len four_adapter:\", len(four_adapter))\n",
    "print(\"len 5p barcode:\", len(fivePBarcode))\n",
    "print(\"len barcode2:\", len(barcode2))\n",
    "print(\"len barcode3:\", len(barcode3))\n",
    "print(\"len barcode4:\", len(barcode4))\n",
    "\n",
    "print(\"len two_adapter + 3p barcode:\", len(two_adapter) + len(barcode2))\n",
    "print(\"len three_adapter + 5p barcode:\", len(three_adapter)+len(fivePBarcode))\n",
    "print(\"len four_adapter + 5p barcode:\", len(four_adapter)+len(fivePBarcode))\n",
    "print(\"total: \", len(two_adapter) + len(barcode2) + len(fivePBarcode))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [],
   "source": [
    "#makes substituted sequence string given seq and rand nuc list\n",
    "def generate_sub_string(seq, rand_nuc_list, rand_subs_list):\n",
    "    seq_list = list(seq)\n",
    "#    print(\"length seq:\", len(seq_list))\n",
    "#    print(\"length rand nuc list\", len(rand_nuc_list))\n",
    "#    print(\"length rand subs list\", len(rand_subs_list))\n",
    "    for i in np.arange(len(rand_nuc_list)):\n",
    "#        print(\"rand_subs_list[\" + str(i) + \"]:\" + str(rand_subs_list[i]))\n",
    "#        print(\"seq_list[rand_subs_list[i]]:\", seq_list[rand_subs_list[i]])\n",
    "#        print(\"rand_nuc_list[i]\", rand_nuc_list[i])\n",
    "        seq_list[rand_subs_list[i]] = rand_nuc_list[i]\n",
    "    sub_str = \"\".join(seq_list)\n",
    "#    print(\"success!! returning. \")\n",
    "#    print()\n",
    "    return sub_str\n",
    "\n",
    "#makes inserted sequence given seq and rand nuc list\n",
    "def generate_insert_string(seq, rand_nuc_list, rand_subs_list):\n",
    "    seq_list = list(seq)\n",
    "    for i in np.arange(len(rand_nuc_list)):\n",
    "        index = rand_subs_list[i] + i\n",
    "        seq_list.insert(index, rand_nuc_list[i])\n",
    "    insert_str = \"\".join(seq_list)\n",
    "    return insert_str\n",
    "\n",
    "#makes deleted sequence given seq and rand nuc list\n",
    "def generate_delete_string(seq, rand_subs_list):\n",
    "    seq_list = list(seq)\n",
    "#    print(\"length seq:\", len(seq_list))\n",
    "#    print(\"length rand subs list\", len(rand_subs_list))\n",
    "    for i in np.arange(int(len(rand_subs_list))):    \n",
    "#        print(\"rand_subs_list[\" + str(i) + \"]:\" + str(rand_subs_list[i]))\n",
    "        index = rand_subs_list[i] -i\n",
    "#        print(\"seq_list[rand_subs_list[i]]:\", seq_list[index])\n",
    "        del seq_list[index]\n",
    "    del_str = \"\".join(seq_list)\n",
    "    return del_str\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_test_strings(barcode, barcodeRC, adapter, adapterRC):    \n",
    "    test_strings = []\n",
    "    #First: exact matches\n",
    "    '''#Ar3r5r\n",
    "    test_ar3r5r = adapterRC + barcodeRC + fivePBarcodeRC\n",
    "    test_strings.append(test_ar3r5r)\n",
    "    #A53A\n",
    "    test_a53a = adapter +\"\".join(rd.choices(nucleotides, k=50))+ fivePBarcode+ \"\".join(rd.choices(nucleotides, k=100)) + barcode+ \"\".join(rd.choices(nucleotides, k=10)) + adapter\n",
    "    test_strings.append(test_a53a)\n",
    "    #5rA3Ar\n",
    "    test_5ra3ar = fivePBarcodeRC + adapter + \"\".join(rd.choices(nucleotides, k=10)) + barcode+\"\".join(rd.choices(nucleotides, k=100)) +adapterRC\n",
    "    test_strings.append(test_5ra3ar)\n",
    "    #start5\n",
    "    test_start5= fivePBarcode + \"\".join(rd.choices(nucleotides, k=(1000)))\n",
    "    test_strings.append(test_start5)\n",
    "    #A_rand_A\n",
    "    test_a_rand_a = adapter + \"\".join(rd.choices(nucleotides, k=900)) + adapter\n",
    "    test_strings.append(test_a_rand_a)\n",
    "    #rand\n",
    "    test_random=\"\".join(rd.choices(nucleotides, k=(1000)))\n",
    "    test_strings.append(test_random)\n",
    "    #end in 3\n",
    "    test_endIn3pBarcode= \"\".join(rd.choices(nucleotides, k=1000)) + barcode\n",
    "    test_strings.append(test_endIn3pBarcode)\n",
    "    '''\n",
    "    #########\n",
    "    #Next: mutated matches\n",
    "    bar2_list = list(barcode)\n",
    "    rand_subs_list = rd.choices(np.arange(len(barcode)), k=int(len(barcode)/10)) #pacbio MUCH more accurate\n",
    "    #here, error in sequencing is 10%\n",
    "    #pacbio has error of about 0.001%\n",
    "    #source: https://www.pacb.com/uncategorized/a-closer-look-at-accuracy-in-pacbio/\n",
    "    \n",
    "    rand_nuc_list = rd.choices(nucleotides, k=len(rand_subs_list))\n",
    "    #3p barcode\n",
    "    rand_subs_list = rd.choices(np.arange(len(barcode)), k=int(len(barcode)/10))\n",
    "    rand_nuc_list = rd.choices(nucleotides, k=len(rand_subs_list))\n",
    "    sub_bar = generate_sub_string(barcode, rand_nuc_list, rand_subs_list)\n",
    "    sub_bar_rc = cf.gen_rev_complement(sub_bar)\n",
    "    insert_bar = generate_insert_string(barcode,rand_nuc_list, rand_subs_list)\n",
    "    insert_bar_rc = cf.gen_rev_complement(insert_bar)\n",
    "    del_bar = generate_delete_string(barcode, rand_subs_list)\n",
    "    del_bar_rc = cf.gen_rev_complement(del_bar)\n",
    "    \n",
    "    #adapter\n",
    "    rand_subs_list = rd.choices(np.arange(len(adapter)), k=int(len(adapter)/10))\n",
    "    rand_nuc_list = rd.choices(nucleotides, k=len(rand_subs_list))\n",
    "    sub_adapter = generate_sub_string(adapter, rand_nuc_list, rand_subs_list)\n",
    "    sub_adapter_rc = cf.gen_rev_complement(sub_adapter)\n",
    "    insert_adapter = generate_insert_string(adapter, rand_nuc_list, rand_subs_list)\n",
    "    insert_adapter_rc = cf.gen_rev_complement(insert_adapter)\n",
    "    del_adapter = generate_delete_string(adapter, rand_subs_list)\n",
    "    del_adapter_rc = cf.gen_rev_complement(del_adapter)\n",
    "    \n",
    "    #5p barcode\n",
    "    rand_subs_list = rd.choices(np.arange(len(fivePBarcode)), k=int(len(fivePBarcode)/10))\n",
    "    rand_nuc_list = rd.choices(nucleotides, k=len(rand_subs_list))\n",
    "    sub_5 = generate_sub_string(fivePBarcode, rand_nuc_list, rand_subs_list)\n",
    "    sub_5_rc = cf.gen_rev_complement(sub_5)\n",
    "    insert_5 = generate_insert_string(fivePBarcode, rand_nuc_list, rand_subs_list)\n",
    "    insert_5_rc = cf.gen_rev_complement(insert_5)\n",
    "    del_5 = generate_delete_string(fivePBarcode, rand_subs_list)\n",
    "    del_5_rc = cf.gen_rev_complement(del_5)\n",
    "    \n",
    "    \n",
    "    #####\n",
    "    #test strings\n",
    "    #####\n",
    "    #Ar3r5r\n",
    "    test_ar3r5r_sub = sub_adapter_rc + sub_bar_rc + sub_5_rc\n",
    "    test_ar3r5r_insert = insert_adapter_rc + insert_bar_rc + insert_5_rc\n",
    "    test_ar3r5r_del = del_adapter_rc + del_bar_rc + del_5_rc\n",
    "    test_strings.append(test_ar3r5r_sub)\n",
    "    test_strings.append(test_ar3r5r_insert)\n",
    "    test_strings.append(test_ar3r5r_del)\n",
    "    #A53A\n",
    "    test_a53a_sub = sub_adapter +\"\".join(rd.choices(nucleotides, k=50))+ sub_5 + \"\".join(rd.choices(nucleotides, k=100)) + sub_bar+ \"\".join(rd.choices(nucleotides, k=10)) + sub_adapter\n",
    "    test_a53a_insert = insert_adapter +\"\".join(rd.choices(nucleotides, k=50))+ insert_5 + \"\".join(rd.choices(nucleotides, k=100)) + insert_bar+ \"\".join(rd.choices(nucleotides, k=10)) + insert_adapter\n",
    "    test_a53a_del = del_adapter +\"\".join(rd.choices(nucleotides, k=50))+ del_5 + \"\".join(rd.choices(nucleotides, k=100)) + del_bar+ \"\".join(rd.choices(nucleotides, k=10)) + del_adapter\n",
    "    test_strings.append(test_a53a_sub)\n",
    "    test_strings.append(test_a53a_insert)\n",
    "    test_strings.append(test_a53a_del)\n",
    "\n",
    "    #5rA3Ar\n",
    "    test_5ra3ar_sub = sub_5_rc + sub_adapter + \"\".join(rd.choices(nucleotides, k=10)) + sub_bar +\"\".join(rd.choices(nucleotides, k=100)) +sub_adapter_rc\n",
    "    test_5ra3ar_insert = insert_5_rc + insert_adapter + \"\".join(rd.choices(nucleotides, k=10)) + insert_bar +\"\".join(rd.choices(nucleotides, k=100)) +insert_adapter_rc\n",
    "    test_5ra3ar_del = del_5_rc + del_adapter + \"\".join(rd.choices(nucleotides, k=10)) + del_bar +\"\".join(rd.choices(nucleotides, k=100)) +del_adapter_rc\n",
    "    test_strings.append(test_5ra3ar_sub)\n",
    "    test_strings.append(test_5ra3ar_insert)\n",
    "    test_strings.append(test_5ra3ar_del)\n",
    "    return test_strings\n",
    "\n",
    "#actual test strings\n",
    "test_2B01_strings = gen_test_strings(barcode2, barcode2RC, two_adapter, two_adapter_RC)\n",
    "test_3C01_strings = gen_test_strings(barcode3, barcode3RC, three_adapter, three_adapter_RC)\n",
    "test_4D01_strings = gen_test_strings(barcode4, barcode4RC, four_adapter, four_adapter_RC)\n",
    "#exact matches: Ar3r5r, A53A, 5rA3Ar, ,start5, A_rand_A, rand, endIn3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [],
   "source": [
    "#given a sequence, and a list of lists of [reference, reference character], \n",
    "#annotates the sequence with the location(s) of the reference(s)\n",
    "#does references and their reverse complements, with the RCs having \"r\" \n",
    "    #appended to the reference character\n",
    "#only includes reference(s) with a score significance of >0.99999999\n",
    "#RETURN: of the references that pass the score significance filter, \n",
    "    #returns fusion of the alignment strings returned by glocal alignment function\n",
    "\n",
    "#parameters\n",
    "    #sequence: string\n",
    "    #ref_list: list of lists\n",
    "        #each entry is [reference(string), reference character]\n",
    "def annotate_seq(sequence, ref_list):\n",
    "    fused_list = ['-']*len(sequence) \n",
    "    for refPair in ref_list:\n",
    "        print(\"ref is\" ,refPair[1])\n",
    "        ref = refPair[0]\n",
    "        ref_char = refPair[1]\n",
    "        refRC = cf.gen_rev_complement(ref)\n",
    "        refRC_char = ref_char + 'r'\n",
    "        ref_list = glocal_alignment(gap_penalty=1, sequence=sequence,\n",
    "                                              reference=ref, score_dict = cf.score_dict,\n",
    "                                     reference_char = ref_char)\n",
    "#        print(\"ref_list:\",ref_list)\n",
    "        ref_scores = []\n",
    "        ref_alignments = []\n",
    "        for pair in ref_list:\n",
    "            ref_scores.append(pair[0])\n",
    "            ref_alignments.append(pair[1])\n",
    "#        print(\"ref_scores\",ref_scores)\n",
    "            \n",
    "        rc_list = glocal_alignment(gap_penalty=1, sequence=sequence,\n",
    "                                              reference=refRC, score_dict = cf.score_dict,\n",
    "                                              reference_char = refRC_char)\n",
    " #       print(\"rc_list\",rc_list)\n",
    "        rc_scores = []\n",
    "        rc_alignments = []\n",
    "        for pair in rc_list:\n",
    "            rc_scores.append(pair[0])\n",
    "            rc_alignments.append(pair[1])\n",
    "#        print(\"rc_scores\",rc_scores)\n",
    "            \n",
    "        ref_scoreSig_list = [cf.calc_score_sig(ref_score) for ref_score in ref_scores]\n",
    "        rc_scoreSig_list = [cf.calc_score_sig(rc_score) for rc_score in rc_scores]\n",
    "        print(\"ref score sigs:\", ref_scoreSig_list)\n",
    "        print(\"ref RC score sigs:\",rc_scoreSig_list)\n",
    "#        print()\n",
    "#        print()\n",
    "        THRESHOLD = 0.999999999 #(9) 9's\n",
    "        for i in np.arange(len(ref_scoreSig_list)):\n",
    "            if ref_scoreSig_list[i]>THRESHOLD:\n",
    "                fuse(fused_list,ref_alignments[i])\n",
    "        for i in np.arange(len(rc_scoreSig_list)):\n",
    "            if rc_scoreSig_list[i]>THRESHOLD:\n",
    "                fuse(fused_list,rc_alignments[i])\n",
    "        \n",
    "#    print(\"fused_list: \",fused_list)\n",
    "    return_str = \"\".join(fused_list)\n",
    "#    print(\"string version of fused_list\\n\", return_str)\n",
    "    return return_str\n",
    "    \n",
    "#print(cf.calc_score_sig(45))\n",
    "#print(cf.calc_score_sig(-3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [],
   "source": [
    "#given current fused list and list of what to fuse, fuses the \"to_fuse\" contents to the fused list\n",
    "def fuse(fused, to_fuse):\n",
    "    len_fused = len(fused)\n",
    "    for i in np.arange(len_fused):\n",
    "        if to_fuse[i] != '-':\n",
    "            if fused[i] != '-':\n",
    "                fused[i] = '!'\n",
    "            else:\n",
    "                fused[i] = to_fuse[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ref is 2\n",
      "ref score sigs: [3.033024161965527e-09]\n",
      "ref RC score sigs: [1.0, 1.0]\n",
      "ref is A\n",
      "ref score sigs: [0.9157761915991026, 0.9157761915991026]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is 5\n",
      "ref score sigs: [0.9954419611263292, 0.9954419611263292]\n",
      "ref RC score sigs: [0.9999999999743453, 0.9999999999743453]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.115787235902589e-09, 1.115787235902589e-09]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is A\n",
      "ref score sigs: [0.9157761915991026]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is 5\n",
      "ref score sigs: [0.9954419611263292]\n",
      "ref RC score sigs: [0.9999999999743453, 0.9999999999743453]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.115787235902589e-09]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is A\n",
      "ref score sigs: [0.16607476979884608, 0.16607476979884608]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is 5\n",
      "ref score sigs: [0.9954419611263292]\n",
      "ref RC score sigs: [0.999999999996528]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [0.0004933956569012966]\n",
      "ref is A\n",
      "ref score sigs: [1.0, 1.0]\n",
      "ref RC score sigs: [0.9157761915991026, 0.9157761915991026, 0.9157761915991026]\n",
      "ref is 5\n",
      "ref score sigs: [0.9999999999743453]\n",
      "ref RC score sigs: [0.9983183459741529, 0.9983183459741529, 0.9983183459741529, 0.9983183459741529]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [0.5953903248083102]\n",
      "ref is A\n",
      "ref score sigs: [1.0, 1.0]\n",
      "ref RC score sigs: [0.9157761915991026]\n",
      "ref is 5\n",
      "ref score sigs: [0.9999999999743453, 0.9999999999743453]\n",
      "ref RC score sigs: [0.9993806957312844]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [0.0013400523616629658]\n",
      "ref is A\n",
      "ref score sigs: [1.0, 1.0]\n",
      "ref RC score sigs: [0.9993806957312844]\n",
      "ref is 5\n",
      "ref score sigs: [0.999999999996528]\n",
      "ref RC score sigs: [0.9997720814673112]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [0.0004933956569012966, 0.0004933956569012966]\n",
      "ref is A\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is 5\n",
      "ref score sigs: [0.9999999999743453]\n",
      "ref RC score sigs: [0.9999999985993009]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [0.99998865014638, 0.99998865014638]\n",
      "ref is A\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is 5\n",
      "ref score sigs: [0.9999999999743453]\n",
      "ref RC score sigs: [0.9999999999743453]\n",
      "\n",
      "ref is 2\n",
      "ref score sigs: [1.0, 1.0]\n",
      "ref RC score sigs: [0.009817666458252721, 0.009817666458252721]\n",
      "ref is A\n",
      "ref score sigs: [1.0]\n",
      "ref RC score sigs: [1.0]\n",
      "ref is 5\n",
      "ref score sigs: [0.999999999996528]\n",
      "ref RC score sigs: [0.999999999810436]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#test it!\n",
    "\n",
    "#test fuse:\n",
    "def test_fuse(filler):\n",
    "    filler = '-'\n",
    "    fused = [filler] * 5\n",
    "    toFuse0 = [filler,filler,filler,'0','0']\n",
    "    fuse(fused,toFuse0)\n",
    "    toFuse1 = [filler,'1','1','1',filler]\n",
    "    fuse(fused, toFuse1)\n",
    "    assert fused == [filler, '1', '1', '1', '0']\n",
    "#test_fuse('-')\n",
    "\n",
    "#test annotate_seq\n",
    "ref_list_2 = [[barcode2,'2'], [two_adapter,'A'],[fivePBarcode,'5'] ]\n",
    "ref_list_3 = [[barcode3,'3'], [three_adapter,'A'], [fivePBarcode,'5']]\n",
    "ref_list_4 = [[barcode4,'4'], [four_adapter,'A'], [fivePBarcode,'5']]\n",
    "def test_annotateseq(ref_list, seq_list):\n",
    "    for seq in seq_list:\n",
    "        annotate_seq(seq, ref_list)\n",
    "#        print(annotate_seq(seq, ref_list))\n",
    "        print()\n",
    "test_annotateseq(ref_list_2, test_2B01_strings)\n",
    "#  0       1     2        3         4       5      6\n",
    "#Ar3r5r, A53A, 5rA3Ar, ,start5, A_rand_A, rand, endIn3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
